### PR TITLE
fix(ivy): make defineComponent tree shakable by Closure Compiler

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: packages/core/src/util.ts [hunk 1]
```diff
@@ -96,3 +96,16 @@ export function stringify(token: any): string {
   const newLineIndex = res.indexOf('\n');
   return newLineIndex === -1 ? res : res.substring(0, newLineIndex);
 }
+
+/**
+ * Convince closure compiler that the wrapped function has no side-effects.
+ *
+ * Closure compiler always assumes that `toString` has no side-effects. We use this quirk to
+ * allow us to execute a function but have closure compiler mark the call as no-side-effects.
+ * It is important that the return value fo the `noSideEffects` function be assigned
```
Comment (matpag) [inline packages/core/src/util.ts]: Typos:
fo -> for
function be assigned -> function should be assigned?

**Comment by alfaproject** [comment]
Why not teach Closure Compiler the `#__PURE__` annotation? TS/Babel (and others?) use this annotation for UglifyJS.

File: packages/core/src/util.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/util.ts]: fixed

File: packages/core/src/render3/definition.ts [hunk 2]
```diff
@@ -292,38 +285,53 @@ export function defineComponent<T>(componentDefinition: {
...
   };
-  const feature = componentDefinition.features;
-  feature && feature.forEach((fn) => fn(def));
+  def._ = noSideEffects(() => {
+    const directiveTypes = componentDefinition.directives !;
+    const feature = componentDefinition.features;
+    const pipeTypes = componentDefinition.pipes !;
+    const animations = componentDefinition.animations;
+    if (animations) {
+      def.data = {animations: animations};
```
Comment (JoostK) [inline packages/core/src/render3/definition.ts]: I'm afraid this will not use the `data` property from the definition parameter anymore.

Comment (mhevery) [reply packages/core/src/render3/definition.ts]: fixed

**Comment by mhevery** [comment]
http://test/OCL:216935620:BASE:217228401:1539646262399:8d0da4d

**Review by IgorMinar — APPROVED** [review]
please change the type for `_`. otherwise lgtm

File: packages/core/src/render3/interfaces/definition.ts [hunk 3]
```diff
@@ -271,6 +271,12 @@ export interface ComponentDef<T> extends DirectiveDef<T> {
    * `PipeDefs`s. The function is necessary to be able to support forward declarations.
    */
   pipeDefs: PipeDefListOrFactory|null;
+
+  /**
+   * Used to store the result of `noSideEffects` function so that it is not removed by closure
+   * compiler. The property should never be read.
+   */
+  readonly _?: any;
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/render3/interfaces/definition.ts]: fixed

**Pull request closed without merging**